### ISSUE TITLE
Deserialization fails with "Buffer underflow" after add a new field (Issue with backward compatibility)

### ISSUE THREAD

**Issue by investr777** [EsotericSoftware/kryo#1012]
Kryo version 5.5.0
When data was serialized and after that a new field was added, deserilization fails with exception: `com.esotericsoftware.kryo.kryo5.io.KryoBufferUnderflowException: Buffer underflow` 

Code example:
```
record Data(
    String id
){}

===============================================
var kryoLocal = ThreadLocal.withInitial(() -> {
    Kryo kryo = new Kryo();
    kryo.setRegistrationRequired(false);
    kryo.register(Data.class);
    kryo.setDefaultSerializer(TaggedFieldSerializer.class);
    return kryo;
});

 var data = new Data("1");

var output = new Output(1024, -1);
kryoLocal.get().writeObject(output, data);
output.flush();
var serializedData = output.toBytes(); // [-126, 49]
output.close();
```

After serialized, new field was added:
```
record Data(
    String id,
   String name
){}

=================

var input = new Input(new byte[] {-126, 49});
var deserializedData = kryoLocal.get().readObject(input, Data.class);
input.close();
```
And eventually an exception is thrown: `com.esotericsoftware.kryo.kryo5.io.KryoBufferUnderflowException: Buffer underflow.`

Also I tried to add @Optional to a new field, it does not work.

**Comment by theigl** [comment]
@investr777: Thanks for the detailed report and sorry for the late answer!

The `RecordSerializer` currently does not have support for backwards/forwards compatibility. It is a limitation in its design that I realized only after it was originally contributed by Oracle.

See [this issue](https://github.com/EsotericSoftware/kryo/issues/884) for a more detailed discussion of the problem.

I have created a [POC](https://github.com/EsotericSoftware/kryo/pull/931) that integrates record serialization back into the default `FieldSerializer` and its subclasses. I ran your test-case against this PR and it passes. it will be part of the next major release, Kryo 6.

I'm afraid that for Kryo 5 there is not much we can do to fix this. If you need support for this right now and can't switch to normal classes for cases where you need compatibility, you could write your own record serializer that uses the same logic as `CompatibleFieldSerializer`.

**Comment by investr777** [comment]
@theigl, thank you so much for your reply. 
Will Kryo 6 and Kryo 5 have backward compatibility? Will Kryo 6 be able to deserialized data, which were serialized by Kryo 5?

**Comment by theigl** [comment]
@investr777: Yes, my goal is that Kryo 6 is backwards compatible (See https://github.com/EsotericSoftware/kryo/wiki/Kryo-v6-Ideas).
